File: src/day04/main.py
def main(input):
    p1 = 0
    p2 = 0
    for line in input.splitlines():
        ((a_lo, a_hi), (b_lo, b_hi)) = [
            [int(x) for x in half.split('-')] for half in line.split(',')]
        if max(a_lo, b_lo) <= min(a_hi, b_hi):
            p2 += 1
            if ((a_lo <= b_lo) and (a_hi >= b_hi)) or ((a_lo >= b_lo) and (a_hi <= b_hi)):
                p1 += 1
    print(f'part 1: {p1}')
    print(f'part 2: {p2}')

File: src/day04/main.exs
lines =
  File.read!("puzzle_input/day04")
  |> String.trim()
  |> String.split(~r/\R/, trim: True)

ranges =
  for line <- lines do
    for half <- String.split(line, ",") do
      for x <- String.split(half, "-") do
        {x, _} = Integer.parse(x)
        x
      end
    end
  end

overlaps =
  for [[a_lo, a_hi], [b_lo, b_hi]] <- ranges,
      max(a_lo, b_lo) <= min(a_hi, b_hi),
      do: (a_lo <= b_lo && a_hi >= b_hi) || (a_lo >= b_lo && a_hi <= b_hi)

IO.puts("part 1: #{Enum.count(overlaps, & &1)}")
IO.puts("part 2: #{Enum.count(overlaps)}")
